### ISSUE TITLE
Crashes, 'comm_open' and function definition error

### ISSUE THREAD

**Issue by tonygon** [root-project/cling#123]
Hello, 
first of all, great project, thanks for all the amazing work you've done.

Now, I get some issues with cling on jupyter.
In all the issues described below, the kernel crashes after running a cell.

I've installed the 3 kernels c++11, c++14 and c++17.
C++17 gives out the error
````
error: invalid value 'c++17' in '-std=c++17'
[ClingKernelApp] ERROR | UNKNOWN MESSAGE TYPE: 'comm_open'
[ClingKernelApp] ERROR | UNKNOWN MESSAGE TYPE: 'comm_open'
````
I guess this is related to mac's LLVM version, or c++17 support. Any workarounds to support c++17?

Using the other kernels results in:
````
$jupyter notebook
[I 13:01:34.197 NotebookApp] Serving notebooks from local directory: 
...
[I 13:01:34.197 NotebookApp] Use Control-C to stop this server and shut down all kernels (twice to skip confirmation).
[C 13:01:34.216 NotebookApp] 
    
    Copy/paste this URL into your browser when you connect for the first time,
    to login with a token:
        http://localhost:8888/?token=760e2f222b86ee6e84814c1c6f712d8568f0f32d6ab5d037
[I 13:01:35.696 NotebookApp] Accepting one-time-token-authenticated connection from ::1
[I 13:02:04.864 NotebookApp] Kernel started: b186650d-400c-424f-afc8-63ccb6c88286
[W 13:02:05.054 NotebookApp] 404 GET /nbextensions/canopy/training/exercise.js?v=20170319130129 (::1) 138.25ms referer=http://localhost:8888/notebooks/cpp_katas/Untitled.ipynb
[W 13:02:05.803 NotebookApp] 404 GET /nbextensions/widgets/notebook/js/extension.js?v=20170319130129 (::1) 3.02ms referer=http://localhost:8888/notebooks/cpp_katas/Untitled.ipynb
[ClingKernelApp] ERROR | UNKNOWN MESSAGE TYPE: 'comm_open'
[ClingKernelApp] ERROR | UNKNOWN MESSAGE TYPE: 'comm_open'
Assertion failed: nbytes == sizeof dummy (bundled/zeromq/src/signaler.cpp:196)
[I 13:02:25.866 NotebookApp] KernelRestarter: restarting kernel (1/5)
WARNING:root:kernel b186650d-400c-424f-afc8-63ccb6c88286 restarted
````

Where a simple function, with includes divided in a previous cell
````
#include <iostream>
#include <vector>
````
````
using namespace std;
int min(vector<int> list) {
  return *(min_element(list.begin(), list.end()));
}
````
Gives this error on the notebook 
````
input_line_4:4:27: error: function definition is not allowed here
int min(vector<int> list) {
                          ^
````
Also, I've noticed this on the console
````
[ClingKernelApp] ERROR | UNKNOWN MESSAGE TYPE: 'comm_open'
Assertion failed: nbytes == sizeof dummy (bundled/zeromq/src/signaler.cpp:196)
````
Several function definitions give this out. Sometimes one that was working, so it's a bit random I think.

As said in the beginning, running a cell, maybe two, crashes the kernel abruptly.

Any ideas?

Thanks!

**Comment by Axel-Naumann** [comment]
I was hoping some other user of the Jupyter kernel had some insight... I'll have to try myself apparently :-(

A bit of additional info, please:
- which platform?
- which version of Jupyter?

And I had to fix the readme; we misspelled the name of the kernels (https://github.com/root-project/cling/commit/7a083b672a85be644c9c3d09021234e3c4dd11af): is that maybe the source of this problem?

**Comment by Djgreene730** [comment]
I'm having the same two issues with my build (Ubuntu 16.04, Jupyter 5.0.0, Python 3.5.2, GCC 5.4.0 20160609). The 'comm_open' issue seams to start when I start a kernel. That doesn't seem to be causing me any trouble, but many functions with output are resulting in the 'Assertion failed' message referenced above.

Here's a very simple example that varies kills it without fail, each line is run in a separate cell:
```
#include<iostream>
std::cout << "Hello world!\n" << std::endl;
.help
__cplusplus
```

In C++11 and C++14, the kernel dies either when printing the output of help or the version info from the macro command, it "seems" random whether it will fail on one or the other,
```
[I 13:07:03.572 NotebookApp] Kernel restarted: 6fe89218-2e71-44fc-985d-e4a90f25da0d
[ClingKernelApp] WARNING | Unknown message type: 'comm_open'
Assertion failed: nbytes == sizeof (dummy) (src/signaler.cpp:303)
[I 13:07:12.573 NotebookApp] KernelRestarter: restarting kernel (1/5)
WARNING:root:kernel 6fe89218-2e71-44fc-985d-e4a90f25da0d restarted
```
Let me know what logs or info I can provide, thanks!

@Axel-Naumann, the misspelling wasn't an issue, it won't install with "c++...", only "cpp..."

**Comment by Axel-Naumann** [comment]
Should be fixed by     8e2700d75380e0e5fab5ef91beb6e360b6a74416! Your example works now, nicely, as it should.
